### PR TITLE
Do not transpile JS packages twice

### PR DESCRIPTION
**Pull request by scinos**
### Summary

Do not transpile some packages at installation time because webpack/jest will transpile them anyway. This allows faster installations (~38% faster) and smaller bundles.

Based on #43679

### Background

Packages in `./packages/` implement a `prepare` script that is run every time we do `yarn install`. For JavaScript packages, this script is using babel to transpile the source to CJS and ESM. Those packages expose the ESM result in `pkg.module` (used by Webpack) and the CJS result in `pkg.main` (used by Node). 

Later, when compiling `./client`, webpack loads the ESM transpiled files, transpiles them again and add them to the bundles.

### Changes

* Do not implement a `prepare` script. Instead, use the `prepack` script to transpile to CJS. This will run the transpilation when `yarn publish` or `yarn pack`, but not on installation (see https://docs.npmjs.com/misc/scripts)

* Change `module` to point to the source code (not the transpiled ESM). This allows distributing the untranspiled source code so consumers of this package (including Calypso) have control over the transpilation options. Specifically, this allows Calypso to transpile them for "evergreen" browsers. Before this change, packages in `./packages` were always transpiled for "fallback" browsers.

* Drop the `prepare` script. This will allow faster `yarn build-packages` (which is run on every `yarn install`). In my machine, in master `yarn build-packages` takes 39.9s to "prepare" 23 packages. After this change, it takes 24.6s (**38% faster**) to prepare 10 packages.

### Testing instructions

* Verify calypso.live loads (both evergreen and fallback) and there are no noticeable errors. If there is something broken entire sections will fail to load, so it should be easy to spot.
* Verify that ICFY reports no size changes (so no new packages are added to the bundle).
* Verify that tests passes (so Jest also understand the changes).

### Update (22/Oct)

- Instead of changing the meaning of `module`, this PR introduces a new `calypso:src` field that points to the untranspiled source code. `pkg.module` and `pkg.main` are unchanged.

### CODE REVIEW

**Comment by scinos** [comment]
Note: this PR was blocked because Jest doesn't know how to resolve packages with a `module` field. Jest provides way to write our own resolver, but its cache won't be clean up on watch mode (which may cause some confusion). See https://github.com/facebook/jest/issues/2702#issuecomment-671947783 for more details.

I raised a [PR](https://github.com/facebook/jest/pull/10393) with the changes required to make the custom resolver work and it just got accepted! As soon as that change is released and we bump Jest to the new version this PR will become unblocked.

**Comment by jsnajdr** [comment]
If we point the `module` field to `src/index.js`, as this patch does for many packages, it means that we start publishing untranspiled sources on NPM. Was that really your intent?

Normally, the CJS and ESM builds differ only in the module format: one uses CJS exports, the other uses the `import` and `export` keywords. The rest of the code is still always transpiled to a widely compatible ES5 syntax, both for CJS and ESM.

After this patch, the code shipped with CJS and ESM will be different.

If we are OK with publishing untranspiled code to NPM, we can remove the package transpilation stuff completely, even for CJS. It's then up to the package consumer to figure out how to consume the sources.

**Comment by scinos** [comment]
> If we point the module field to src/index.js, as this patch does for many packages, it means that we start publishing untranspiled sources on NPM. Was that really your intent?

IMO that's a good thing. We continue to publish a ES5 version, packed as CJS in `main`. This is how consumers can use our code in a way that JustWorks™️. We also publish a modern, untranspiled ES2020 version packed as ESM in `module` for advanced users.

I think that has more value than shipping transpiled ES5 in CJS and ESM.

**Comment by scinos** [comment]
An alternative would be to introduce a third property (`esnext`?) that points to the original source. We would have `main` with ES5+CJS, `module` with ES5+ESM, `esnext` with ES2020+ESM.

**Comment by jsnajdr** [comment]
> An alternative would be to introduce a third property (`esnext`?)

Yeah, that's `calypso:main`, only with a different name.

**Comment by jsnajdr** [comment]
> We continue to publish a ES5 version, packed as CJS in `main`. This is how consumers can use our code in a way that JustWorks.

That sounds good only until one starts to look more deeply at who these CJS consumers are and how they consume the package in practice.

If the consumer uses webpack, then webpack always looks at `module` field before `main`. For these consumers, the CJS version is practically invisible. The consumer will always use the `module` thing, which is newly untranspiled.

The change from transpiled to untranspiled requires caution on the consumer side. E.g., Jetpack uses `i18n-calypso` and it _doesn't_ transpile it (that's specified in the webpack config). Just switching to the newly published version would lead to bundling untranspiled code into the output, including JSX, which would break a lot of things.

The only place where I can see the CJS version really used is `require( 'a8c-package' )` in Node.js. But that's already kind of legacy -- if I write `import from 'a8c-package'` in latest Node.js, I'll get the untranspiled module pointed to by the `module` field, right? And that code will often contain JSX, causing a syntax error.

To summarize, I think there is a mismatch between what our packages expose and how the consumers actually consume it. That mismatch exists even without the changes in this PR -- the PR only highlights them.

**Comment by scinos** [comment]
> If the consumer uses webpack, then webpack always looks at module field before main. For these consumers, the CJS version is practically invisible.

That's the default behaviour, but I'd say that's the consumer's responsibility to change if they are not ok with it. I do agree most people won't change it so for them, `module` is the "main" way to consume the package.

> Just switching to the newly published version would lead to bundling untranspiled code into the output, including JSX, which would break a lot of things.

For the public packages, we could bump the major version number and document the breaking change.

> if I write import from 'a8c-package' in latest Node.js, I'll get the untranspiled module pointed to by the module field, right?

AFAIK, node always ignores the `module` field. You will get `main` or `exports` (if defined).


I think we have three options:

1)  Ship `module` with untranspiled+ESM (aka this PR). It is arguable the simplest option but will require consumers to adapt their webpack settings to also transpile these packages.
2) Ship a new `esnext`, `calypso:main` or any other bespoke name. Adds the complexity of having to remember what that custom field means, but it won't require changes in the consumers if they don't want to. Only those who want to get the benefit of the untranspiled code (eg: Calypso) will have to change.
3) Go with the new conditional exports. It is more future-proof and the way we will eventually go with, but at the moment there is no support in Webpack 4 and no standard way of exposing untranspiled code (webpack proposed a `browser` field).

Sounds like (2) is the most sensible approach right now. We just need a name. `esnext` may get confused with the old `esnext` (https://2ality.com/2017/06/pkg-esnext.html), `calypso:main` is too coupled with Calypso IMO. What about just `src`? It kind of suggests that it points to the original source, not transpiled in any way (may include TS, JSX, ES.next features...).

**Comment by jsnajdr** [comment]
We can also specify the requirements by consumers and then work from there.

For browsers, I think the only practical consumer are bundlers, and to these, we can simply ship the untranspiled sources (even TypeScript) and shift the responsibility entirely to the consumer, i.e., the webpack config author. It's the consumer, not the package publisher, who knows what their target browsers and polyfills are. Transpilation by publishers only leads to suboptimal code (i.e., transpiled too much).

Then there's the Node.js runtime that imports the package at runtime using the Node.js resolver and loader. Node.js doesn't understand JSX, so _some_ transpilation by publisher will always need to happen. (Here the typical package is a React component that's also used for SSR in Node.)

For Node.js without ESM and `exports`, we need a CJS export pointed to by `main` field, with transpiled JSX and reasonably stable JS syntax.

For latest Node.js with ESM unflagged, we need an ESM export with the same code content as CJS.

Can all that be encoded in `package.json` in such a way that each consumer picks up the correct thing? I don't know at this moment.

> Go with the new conditional exports. It is more future-proof and the way we will eventually go with, but at the moment there is no support in Webpack 4

Is there support in Webpack 5? We can write a webpack plugin for `exports`, that shouldn't be hard.

> no standard way of exposing untranspiled code

This standardization debate has been going on for years, with no results. I'm afraid we'll need to go with something that's not standard, but is reasonably well designed and has a chance for wider adoption.

> `calypso:main` is too coupled with Calypso IMO. What about just `src`?

[This article](https://dev.to/garylchew/bringing-modern-javascript-to-libraries-432c) proposes `browser2017`.

**Comment by jsnajdr** [comment]
When exposing the different exports from a package, my experience from #43679 has been that the biggest bottleneck is the TypeScript compiler. It has its own resolver which is not configurable. It's possible to write your own resolver, but that involves implementing the "compiler host" interface and somehow wiring it all together.

It affects JS-only packages, too, as long as a TS source imports them. The TS compiler will look for the `index.d.ts` type definitions, the `types` field... TS needs to be treated as another common consumer that needs to get the right thing when importing a JS package.

**Comment by scinos** [comment]
> calypso:main is too coupled with Calypso IMO. What about just src?

> This article proposes browser2017.

The article also proposes `esnext` (equivalent to webpack's conditional `browser` proposal) as an alternative.
I don't think any of those applies, because those are for exposing transpiled code. I think we should introduce `src` as a way to expose untranspiled code

@jsnajdr In your opinion, is the name of the export the only thing that needs to be changed to get this merged to master? Or is there something else?

**Comment by jsnajdr** [comment]
> In your opinion, is the name of the export the only thing that needs to be changed to get this merged to master?

I'd like to discuss and design the package-export strategy as a whole before starting to ship code. Right now, I'm working on understanding the conditional exports proposal more deeply. These are big changes and I think that trying to merge this or #43679 as expediently as possible is not the right focus.

**Comment by jsnajdr** [comment]
It seems that the current patch is able to satisfy the requirements in https://github.com/Automattic/wp-calypso/pull/44824#issuecomment-674719849:

The `module` field points to the untranspiled sources, and it is what is picked by bundlers to further transpile and process. Node.js reportedly never looks at the `module` field: it only looks at `main` or the new `exports`.

When loaded natively by Node.js, the `dist/cjs` folder pointed to by `main` field is used. That makes the package consumable by Node.js apps.

No custom field (`esnext` or `calypso:main`) is strictly needed here.

Open questions:
- the `dist/esm` folder seems to be redundant now. There is no use case for it.
- can TypeScript compiler always load the affected packages' types?
- there are packages/subpackages that export different files for browsers and for Node.js, using the `browser` field. Are we compatible with that?

**Comment by scinos** [comment]
> * the `dist/esm` folder seems to be redundant now. There is no use case for it.

Correct, we can delete it.

> * can TypeScript compiler always load the affected packages' types?

I'll test.

> * there are packages/subpackages that export different files for browsers and for Node.js, using the `browser` field. Are we compatible with that?

Yes

File: test/server/jest.config.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ module.exports = {
 	rootDir: '../../client',
 	roots: [ '<rootDir>/server' ],
 	testEnvironment: 'node',
+	resolver: '<rootDir>../test/module-resolver.js',
```
[the remark this answers is not in the document]
Comment (scinos) [reply test/server/jest.config.js]: Done

File: client/webpack.config.node.js [hunk 2]
```diff
@@ -56,6 +57,15 @@ function getExternals() {
...
 				/[^/]?wp-calypso-client\//,
+
+				...fs
+					.readdirSync( path.join( __dirname, '../packages' ), { withFileTypes: true } )
+					.filter( ( file ) => file.isDirectory() )
+					.map( ( file ) =>
+						require( path.join( __dirname, '../packages', file.name, 'package.json' ) )
+					)
+					.filter( ( pkg ) => pkg.module )
+					.map( ( pkg ) => pkg.name ),
```
[the remark this answers is not in the document]
Comment (scinos) [reply client/webpack.config.node.js:64]: Done

**Comment by scinos** [comment]
I'm reviving this PR, as it is still beneficial for dev speed (avoid transpiling many packages twice) and client performance (allow transpilation for evergreen).

I'm reintroducing a custom field (`calypso:src`) to expose the untranspiled code. Pros of this approach:

- Does not change the semantics of the existing `module`, so consumers of these packages don't need to change anything in their bundler config and their build times won't get affected.
- Namespaced to `calypso:` to avoid potential conflicts with existing tools or future changes to `package.json` spec.
- Leaves a escape hatch in case some tool doesn't understand `calypso:src`. We can always run `yarn build` to generate `module`/`main` and then run that tool.

Cons:
- Consumers will still get the suboptimal behaviour of double transpilation if they continue to use `module` or `main` (once transpiled by us, a second one likely transpiled by them). As we may target a different set of browsers/node runtimes in our transpilation, the resulting code is not as optimized as it could be. It is worth mentioning that this is the existing situation today anyway.
- Adds yet another "calypsoism" that calypso devs will have to understand.
 
After considering pros and cons, I think not _requiring_ every consumer to change their webpack config in a non-trivial way (and instead allow them to opt-in if they want) is critical.

Pending questions:

> can TypeScript compiler always load the affected packages' types?

Yes it does. I was able to compile evergreen,  fallback and every TS package without errors. Also VSCode seems to correctly pick types even when the package has `calypso:src`.

> there are packages/subpackages that export different files for browsers and for Node.js, using the browser field. Are we compatible with that?

Yes, it still works. It is actually used by `packages/wpcom.js` and if it wasn't working, webpack wouldn't be able to compile evergreen or fallback.

**Comment by jsnajdr** [comment]
@scinos How do you think this PR is related to #43679 which has a similar goal? They both now use `calypso:main` or `calypso:src`. Is there a big overlap between them? Are they at odds with each other?

I didn't look at this problem in the last few months, so I forgot almost everything I knew about both PRs 🙂

**Comment by scinos** [comment]
I think this PR is a subset of  #43679.  #43679 also changes TypeScript projects and bundles all node_modules into server bundler.

**Comment by jsnajdr** [comment]
> bundles all node_modules into server bundler.

Yeah, that's superseded by the `packagesInMonorepo` test, also present in #45774 in a slightly different form.

Thanks for reviving this effort -- it's an important improvement that's been dormant for a few months unfortunately 🙁

**Comment by scinos** [comment]
Strange, I expected all bundles to be smaller as they are transpiled for evergreen now, but somehow some sections got bigger 🤨 . I'll investigate

**Comment by scinos** [comment]
> Strange, I expected all bundles to be smaller as they are transpiled for evergreen now

False alarm. Changes in chunk sizes are caused by webpack heuristics moving things around, but overall, this PR makes evergreen build ~30kb smaller (parsed size, from http://iscalypsofastyet.com/p/branch?branch=try/dont-transpile-packages). Manually inspecting a few compiled files I can see the diff comes, as expected, to a more effective transpilation (eg: there are no more `_babel_runtime_helpers_objectSpread2` helpers)

**Comment by sirbrillig** [comment]
Just FYI that I created a new package, `shopping-cart`, in https://github.com/Automattic/wp-calypso/pull/46532, in case that also needs to be updated here.

**Comment by scinos** [comment]
@sirbrillig Thank you for letting me know :) I'll include it in this PR.

File: packages/babel-plugin-transform-wpcalypso-async/jest.config.js [hunk 3]
```diff
@@ -2,4 +2,5 @@ module.exports = {
 	preset: '@automattic/calypso-build',
 	rootDir: __dirname,
 	cacheDirectory: '<rootDir>/../../.cache/jest',
+	resolver: '<rootDir>/../../test/module-resolver.js',
```
Comment (jsnajdr) [inline packages/babel-plugin-transform-wpcalypso-async/jest.config.js:5]: This would be nice to move to some shared preset. The packages' Jest configs are getting repetitive: the `cacheDirectory` is always the same, and now the `resolver` field is, too.

By the way, even the `rootDir` field could be removed one day, when [this Jest bug](https://github.com/facebook/jest/issues/8293) gets fixed.

File: test/module-resolver.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * @param {string} request The package being requested
+ * @param {*} options Options for the resolver
+ */
+module.exports = ( request, options ) => {
+	return options.defaultResolver( request, {
+		...options,
+		packageFilter: ( pkg ) => {
+			//TODO: when Jest moves to resolver v2, this method will receive a second argument that points to the real package file (ie: no symlink)
+			//We can use it to determine if the package file is within the repo
+			if ( packages.includes( pkg.name ) ) {
```
Comment (jsnajdr) [inline test/module-resolver.js:34]: Is this filtering by name even needed? What if we just checked if `calypso:src` field is present in `package.json` and returned a patched `main` if it is?

**Review by jsnajdr — COMMENTED** [review]
@scinos would you mind if I pushed a few little changes to this branch? Locally, I did a rebase to latest `master` and removed the `packagesInMonorepo` code, as the same thing has been merged this week together with #45774. I guess that I could just push it when I already did the work locally, in order to test the PR.

To confirm my understanding: this PR adds `calypso:src` and skips build for packages that are written in standard JS and transpiled using Babel. It keeps TypeScript packages unchanged, mainly because of the troubles with configuring a custom resolver for TypeScript compiler. In #43679 I tried to remove the build step for TypeScript packages, too, but that works only when they are compiled with Babel (with TS plugin). `tsc` doesn't find the imported modules.

File: packages/babel-plugin-transform-wpcalypso-async/jest.config.js (hunk 3 above)
Comment (scinos) [reply packages/babel-plugin-transform-wpcalypso-async/jest.config.js:5]: I couldn't figure out how to do it. When I added the `cacheDirectory` to the base jest config I got some problems relative to how jest resolves paths... I don't remember the details, I'll try again.

File: test/module-resolver.js (hunk 4 above)
Comment (scinos) [reply test/module-resolver.js:34]: Good point, probably not needed, I'll remove it.

File: packages/babel-plugin-transform-wpcalypso-async/jest.config.js (hunk 3 above)
Comment (scinos) [reply packages/babel-plugin-transform-wpcalypso-async/jest.config.js:5]: Created a separate PR for it: #46953 46953

**Comment by scinos** [comment]
> @scinos would you mind if I pushed a few little changes to this branch?

Of course I don't mind! :) Feel free to improve it.

> did a rebase to latest master and removed the packagesInMonorepo code

I've made those changes as well.

> To confirm my understanding: this PR adds `calypso:src` and skips build for packages that are written in standard JS and transpiled using Babel. It keeps TypeScript packages unchanged, mainly because of the troubles with configuring a custom resolver for TypeScript compiler

Correct.

**Comment by jsnajdr** [comment]
> I've made those changes as well.

Yeah, you already made the changes I had in mind 🙂 I have nothing further to add.

**Review by jsnajdr — APPROVED** [review]
Looks good, including the Jest config sub-PR.

In a related PR, I'm trying to reduce the TypeScript busywork in #47002.
